Using the ElasticsearchAsyncClient with BinaryData and a complex object fails on derserialization.

### Java API client version

7.17.10

### Java version

11

### Elasticsearch Version

7.17.10

### Problem description

When using the ElasticsearchAsyncClient with BinaryData and a complex object will fail on deserialization. The object can be indexed properly but cannot be retrieved as BinaryData since deserialization will fail. I have modified one of the basic tests included in the elasticsearch-java project as a reproducer.

` @Test
    public void testBinaryDataIngestion() throws Exception {
        String index = "binary-ingestion-test";
        String id = "foo-bar";

        BinaryData data = BinaryData.of("{\"id\":\"27082ce1-d9ab-404e-a810-f2894640edf4\",\"firstName\":\"Jesse\",\"lastName\":\"Pinkman\",\"addresses\":[{\"street\":\"1001 Central Ave NE\",\"city\":\"Albuquerque\",\"state\":\"NM\",\"zip\":\"87106\"}]}".getBytes(), ContentType.APPLICATION_JSON);

        esAsyncClient.index(i -> i
                .index(index)
                .id(id)
                .document(data)
                .refresh(Refresh.True)
        ).get();

        GetResponse<BinaryData> getResponse = esAsyncClient.get(g -> g
                                                                 .index(index)
                                                                 .id(id)
                , BinaryData.class
        ).get();

        Assertions.assertEquals(id, getResponse.id());
        Assertions.assertEquals(
                "{\"id\":\"27082ce1-d9ab-404e-a810-f2894640edf4\",\"firstName\":\"Jesse\",\"lastName\":\"Pinkman\",\"addresses\":[{\"street\":\"1001 Central Ave NE\",\"city\":\"Albuquerque\",\"state\":\"NM\",\"zip\":\"87106\"}]}",
                new String(getResponse.source().asByteBuffer().array(), StandardCharsets.UTF_8)
        );
    }`
    
    The error I receive 
    
    `co.elastic.clients.json.JsonpMappingException: Error deserializing co.elastic.clients.elasticsearch.core.GetResponse: jakarta.json.stream.JsonGenerationException: Can not start an object, expecting field name (context: Object) (JSON path: _source) (line no=1, column no=232, offset=-1)
java.util.concurrent.ExecutionException: co.elastic.clients.json.JsonpMappingException: Error deserializing co.elastic.clients.elasticsearch.core.GetResponse: jakarta.json.stream.JsonGenerationException: Can not start an object, expecting field name (context: Object) (JSON path: _source) (line no=1, column no=232, offset=-1)
	at java.base/java.util.concurrent.CompletableFuture.reportGet(CompletableFuture.java:395)
	at java.base/java.util.concurrent.CompletableFuture.get(CompletableFuture.java:1999)
	at org.kinotic.structures.entity.EntityCrudTests.testBinaryDataIngestion(EntityCrudTests.java:129)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:566)
	at org.junit.platform.commons.util.ReflectionUtils.invokeMethod(ReflectionUtils.java:725)
	at org.junit.jupiter.engine.execution.MethodInvocation.proceed(MethodInvocation.java:60)
	at org.junit.jupiter.engine.execution.InvocationInterceptorChain$ValidatingInvocation.proceed(InvocationInterceptorChain.java:131)
	at org.junit.jupiter.engine.extension.TimeoutExtension.intercept(TimeoutExtension.java:149)
	at org.junit.jupiter.engine.extension.TimeoutExtension.interceptTestableMethod(TimeoutExtension.java:140)
	at org.junit.jupiter.engine.extension.TimeoutExtension.interceptTestMethod(TimeoutExtension.java:84)
	at org.junit.jupiter.engine.execution.ExecutableInvoker$ReflectiveInterceptorCall.lambda$ofVoidMethod$0(ExecutableInvoker.java:115)
	at org.junit.jupiter.engine.execution.ExecutableInvoker.lambda$invoke$0(ExecutableInvoker.java:105)
	at org.junit.jupiter.engine.execution.InvocationInterceptorChain$InterceptedInvocation.proceed(InvocationInterceptorChain.java:106)
	at org.junit.jupiter.engine.execution.InvocationInterceptorChain.proceed(InvocationInterceptorChain.java:64)
	at org.junit.jupiter.engine.execution.InvocationInterceptorChain.chainAndInvoke(InvocationInterceptorChain.java:45)
	at org.junit.jupiter.engine.execution.InvocationInterceptorChain.invoke(InvocationInterceptorChain.java:37)
	at org.junit.jupiter.engine.execution.ExecutableInvoker.invoke(ExecutableInvoker.java:104)
	at org.junit.jupiter.engine.execution.ExecutableInvoker.invoke(ExecutableInvoker.java:98)
	at org.junit.jupiter.engine.descriptor.TestMethodTestDescriptor.lambda$invokeTestMethod$7(TestMethodTestDescriptor.java:214)
	at org.junit.platform.engine.support.hierarchical.ThrowableCollector.execute(ThrowableCollector.java:73)
	at org.junit.jupiter.engine.descriptor.TestMethodTestDescriptor.invokeTestMethod(TestMethodTestDescriptor.java:210)
	at org.junit.jupiter.engine.descriptor.TestMethodTestDescriptor.execute(TestMethodTestDescriptor.java:135)
	at org.junit.jupiter.engine.descriptor.TestMethodTestDescriptor.execute(TestMethodTestDescriptor.java:66)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$6(NodeTestTask.java:151)
	at org.junit.platform.engine.support.hierarchical.ThrowableCollector.execute(ThrowableCollector.java:73)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$8(NodeTestTask.java:141)
	at org.junit.platform.engine.support.hierarchical.Node.around(Node.java:137)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$9(NodeTestTask.java:139)
	at org.junit.platform.engine.support.hierarchical.ThrowableCollector.execute(ThrowableCollector.java:73)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.executeRecursively(NodeTestTask.java:138)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.execute(NodeTestTask.java:95)
	at java.base/java.util.ArrayList.forEach(ArrayList.java:1541)
	at org.junit.platform.engine.support.hierarchical.SameThreadHierarchicalTestExecutorService.invokeAll(SameThreadHierarchicalTestExecutorService.java:41)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$6(NodeTestTask.java:155)
	at org.junit.platform.engine.support.hierarchical.ThrowableCollector.execute(ThrowableCollector.java:73)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$8(NodeTestTask.java:141)
	at org.junit.platform.engine.support.hierarchical.Node.around(Node.java:137)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$9(NodeTestTask.java:139)
	at org.junit.platform.engine.support.hierarchical.ThrowableCollector.execute(ThrowableCollector.java:73)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.executeRecursively(NodeTestTask.java:138)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.execute(NodeTestTask.java:95)
	at java.base/java.util.ArrayList.forEach(ArrayList.java:1541)
	at org.junit.platform.engine.support.hierarchical.SameThreadHierarchicalTestExecutorService.invokeAll(SameThreadHierarchicalTestExecutorService.java:41)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$6(NodeTestTask.java:155)
	at org.junit.platform.engine.support.hierarchical.ThrowableCollector.execute(ThrowableCollector.java:73)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$8(NodeTestTask.java:141)
	at org.junit.platform.engine.support.hierarchical.Node.around(Node.java:137)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$9(NodeTestTask.java:139)
	at org.junit.platform.engine.support.hierarchical.ThrowableCollector.execute(ThrowableCollector.java:73)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.executeRecursively(NodeTestTask.java:138)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.execute(NodeTestTask.java:95)
	at org.junit.platform.engine.support.hierarchical.SameThreadHierarchicalTestExecutorService.submit(SameThreadHierarchicalTestExecutorService.java:35)
	at org.junit.platform.engine.support.hierarchical.HierarchicalTestExecutor.execute(HierarchicalTestExecutor.java:57)
	at org.junit.platform.engine.support.hierarchical.HierarchicalTestEngine.execute(HierarchicalTestEngine.java:54)
	at org.junit.platform.launcher.core.EngineExecutionOrchestrator.execute(EngineExecutionOrchestrator.java:108)
	at org.junit.platform.launcher.core.EngineExecutionOrchestrator.execute(EngineExecutionOrchestrator.java:88)
	at org.junit.platform.launcher.core.EngineExecutionOrchestrator.lambda$execute$0(EngineExecutionOrchestrator.java:54)
	at org.junit.platform.launcher.core.EngineExecutionOrchestrator.withInterceptedStreams(EngineExecutionOrchestrator.java:67)
	at org.junit.platform.launcher.core.EngineExecutionOrchestrator.execute(EngineExecutionOrchestrator.java:52)
	at org.junit.platform.launcher.core.DefaultLauncher.execute(DefaultLauncher.java:96)
	at org.junit.platform.launcher.core.DefaultLauncher.execute(DefaultLauncher.java:75)
	at org.gradle.api.internal.tasks.testing.junitplatform.JUnitPlatformTestClassProcessor$CollectAllTestClassesExecutor.processAllTestClasses(JUnitPlatformTestClassProcessor.java:99)
	at org.gradle.api.internal.tasks.testing.junitplatform.JUnitPlatformTestClassProcessor$CollectAllTestClassesExecutor.access$000(JUnitPlatformTestClassProcessor.java:79)
	at org.gradle.api.internal.tasks.testing.junitplatform.JUnitPlatformTestClassProcessor.stop(JUnitPlatformTestClassProcessor.java:75)
	at org.gradle.api.internal.tasks.testing.SuiteTestClassProcessor.stop(SuiteTestClassProcessor.java:61)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:566)
	at org.gradle.internal.dispatch.ReflectionDispatch.dispatch(ReflectionDispatch.java:36)
	at org.gradle.internal.dispatch.ReflectionDispatch.dispatch(ReflectionDispatch.java:24)
	at org.gradle.internal.dispatch.ContextClassLoaderDispatch.dispatch(ContextClassLoaderDispatch.java:33)
	at org.gradle.internal.dispatch.ProxyDispatchAdapter$DispatchingInvocationHandler.invoke(ProxyDispatchAdapter.java:94)
	at com.sun.proxy.$Proxy2.stop(Unknown Source)
	at org.gradle.api.internal.tasks.testing.worker.TestWorker$3.run(TestWorker.java:193)
	at org.gradle.api.internal.tasks.testing.worker.TestWorker.executeAndMaintainThreadName(TestWorker.java:129)
	at org.gradle.api.internal.tasks.testing.worker.TestWorker.execute(TestWorker.java:100)
	at org.gradle.api.internal.tasks.testing.worker.TestWorker.execute(TestWorker.java:60)
	at org.gradle.process.internal.worker.child.ActionExecutionWorker.execute(ActionExecutionWorker.java:56)
	at org.gradle.process.internal.worker.child.SystemApplicationClassLoaderWorker.call(SystemApplicationClassLoaderWorker.java:133)
	at org.gradle.process.internal.worker.child.SystemApplicationClassLoaderWorker.call(SystemApplicationClassLoaderWorker.java:71)
	at worker.org.gradle.process.internal.worker.GradleWorkerMain.run(GradleWorkerMain.java:69)
	at worker.org.gradle.process.internal.worker.GradleWorkerMain.main(GradleWorkerMain.java:74)
Caused by: co.elastic.clients.json.JsonpMappingException: Error deserializing co.elastic.clients.elasticsearch.core.GetResponse: jakarta.json.stream.JsonGenerationException: Can not start an object, expecting field name (context: Object) (JSON path: _source) (line no=1, column no=232, offset=-1)
	at app//co.elastic.clients.json.JsonpMappingException.from0(JsonpMappingException.java:134)
	at app//co.elastic.clients.json.JsonpMappingException.from(JsonpMappingException.java:121)
	at app//co.elastic.clients.json.ObjectDeserializer.deserialize(ObjectDeserializer.java:218)
	at app//co.elastic.clients.json.ObjectDeserializer.deserialize(ObjectDeserializer.java:148)
	at app//co.elastic.clients.json.JsonpDeserializer.deserialize(JsonpDeserializer.java:77)
	at app//co.elastic.clients.json.ObjectBuilderDeserializer.deserialize(ObjectBuilderDeserializer.java:79)
	at app//co.elastic.clients.json.DelegatingDeserializer$SameType.deserialize(DelegatingDeserializer.java:43)
	at app//co.elastic.clients.json.DelegatingDeserializer$SameType.deserialize(DelegatingDeserializer.java:43)
	at app//co.elastic.clients.transport.endpoints.EndpointWithResponseMapperAttr$1.deserialize(EndpointWithResponseMapperAttr.java:56)
	at app//co.elastic.clients.transport.rest_client.RestClientTransport.decodeResponse(RestClientTransport.java:377)
	at app//co.elastic.clients.transport.rest_client.RestClientTransport.getHighLevelResponse(RestClientTransport.java:347)
	at app//co.elastic.clients.transport.rest_client.RestClientTransport.access$200(RestClientTransport.java:70)
	at app//co.elastic.clients.transport.rest_client.RestClientTransport$1.onSuccess(RestClientTransport.java:181)
	at app//org.elasticsearch.client.RestClient$FailureTrackingResponseListener.onSuccess(RestClient.java:678)
	at app//org.elasticsearch.client.RestClient$1.completed(RestClient.java:399)
	at app//org.elasticsearch.client.RestClient$1.completed(RestClient.java:393)
	at app//org.apache.http.concurrent.BasicFuture.completed(BasicFuture.java:122)
	at app//org.apache.http.impl.nio.client.DefaultClientExchangeHandlerImpl.responseCompleted(DefaultClientExchangeHandlerImpl.java:182)
	at app//org.apache.http.nio.protocol.HttpAsyncRequestExecutor.processResponse(HttpAsyncRequestExecutor.java:448)
	at app//org.apache.http.nio.protocol.HttpAsyncRequestExecutor.inputReady(HttpAsyncRequestExecutor.java:338)
	at app//org.apache.http.impl.nio.DefaultNHttpClientConnection.consumeInput(DefaultNHttpClientConnection.java:265)
	at app//org.apache.http.impl.nio.client.InternalIODispatch.onInputReady(InternalIODispatch.java:87)
	at app//org.apache.http.impl.nio.client.InternalIODispatch.onInputReady(InternalIODispatch.java:40)
	at app//org.apache.http.impl.nio.reactor.AbstractIODispatch.inputReady(AbstractIODispatch.java:114)
	at app//org.apache.http.impl.nio.reactor.BaseIOReactor.readable(BaseIOReactor.java:162)
	at app//org.apache.http.impl.nio.reactor.AbstractIOReactor.processEvent(AbstractIOReactor.java:337)
	at app//org.apache.http.impl.nio.reactor.AbstractIOReactor.processEvents(AbstractIOReactor.java:315)
	at app//org.apache.http.impl.nio.reactor.AbstractIOReactor.execute(AbstractIOReactor.java:276)
	at app//org.apache.http.impl.nio.reactor.BaseIOReactor.execute(BaseIOReactor.java:104)
	at app//org.apache.http.impl.nio.reactor.AbstractMultiworkerIOReactor$Worker.run(AbstractMultiworkerIOReactor.java:591)
	at java.base@11.0.14.1/java.lang.Thread.run(Thread.java:829)
Caused by: jakarta.json.stream.JsonGenerationException: Can not start an object, expecting field name (context: Object)
	at app//co.elastic.clients.json.jackson.JacksonUtils.convertException(JacksonUtils.java:32)
	at app//co.elastic.clients.json.jackson.JacksonJsonpGenerator.writeStartObject(JacksonJsonpGenerator.java:57)
	at app//co.elastic.clients.json.JsonpUtils.copy(JsonpUtils.java:160)
	at app//co.elastic.clients.json.JsonpUtils.copy(JsonpUtils.java:173)
	at app//co.elastic.clients.json.JsonpUtils.copy(JsonpUtils.java:164)
	at app//co.elastic.clients.util.ByteArrayBinaryData$Deserializer.deserialize(ByteArrayBinaryData.java:89)
	at app//co.elastic.clients.util.ByteArrayBinaryData$Deserializer.deserialize(ByteArrayBinaryData.java:79)
	at app//co.elastic.clients.util.BinaryData$1.deserialize(BinaryData.java:99)
	at app//co.elastic.clients.util.BinaryData$1.deserialize(BinaryData.java:96)
	at app//co.elastic.clients.json.JsonpDeserializer.deserialize(JsonpDeserializer.java:77)
	at app//co.elastic.clients.json.NamedDeserializer.deserialize(NamedDeserializer.java:64)
	at app//co.elastic.clients.json.ObjectDeserializer$FieldObjectDeserializer.deserialize(ObjectDeserializer.java:78)
	at app//co.elastic.clients.json.ObjectDeserializer.deserialize(ObjectDeserializer.java:192)
	... 28 more
Caused by: com.fasterxml.jackson.core.JsonGenerationException: Can not start an object, expecting field name (context: Object)
	at app//com.fasterxml.jackson.core.JsonGenerator._reportError(JsonGenerator.java:2710)
	at app//com.fasterxml.jackson.core.json.JsonGeneratorImpl._reportCantWriteValueExpectName(JsonGeneratorImpl.java:248)
	at app//com.fasterxml.jackson.core.json.UTF8JsonGenerator._verifyValueWrite(UTF8JsonGenerator.java:1166)
	at app//com.fasterxml.jackson.core.json.UTF8JsonGenerator.writeStartObject(UTF8JsonGenerator.java:374)
	at app//co.elastic.clients.json.jackson.JacksonJsonpGenerator.writeStartObject(JacksonJsonpGenerator.java:55)
	... 39 more`

--- a/java-client/src/main/java/co/elastic/clients/json/JsonpUtils.java
+++ b/java-client/src/main/java/co/elastic/clients/json/JsonpUtils.java
@@ -175,7 +175,6 @@ public class JsonpUtils {
 
             case START_ARRAY:
                 generator.writeStartArray();
-                generator.writeStartObject();
                 while ((event = parser.next()) != Event.END_ARRAY) {
                     copy(parser, generator, event);
                 }

--- a/java-client/src/test/java/co/elastic/clients/elasticsearch/ElasticsearchTestServer.java
+++ b/java-client/src/test/java/co/elastic/clients/elasticsearch/ElasticsearchTestServer.java
@@ -77,7 +77,7 @@ public class ElasticsearchTestServer implements AutoCloseable {
     }
 
     public synchronized ElasticsearchTestServer start() {
-        Version version = Version.VERSION.major() < 8 ? new Version(7,17,5,false) : new Version(8,3,3,false);
+        Version version = Version.VERSION.major() < 8 ? new Version(7,17,10,false) : new Version(8,3,3,false);
 
         // Note we could use version.major() + "." + version.minor() + "-SNAPSHOT" but plugins won't install on a snapshot version
         String esImage = "docker.elastic.co/elasticsearch/elasticsearch:" + version;

--- a/java-client/src/test/java/co/elastic/clients/elasticsearch/spec_issues/SpecIssuesTest.java
+++ b/java-client/src/test/java/co/elastic/clients/elasticsearch/spec_issues/SpecIssuesTest.java
@@ -20,15 +20,18 @@
 package co.elastic.clients.elasticsearch.spec_issues;
 
 import co.elastic.clients.documentation.usage.Product;
+import co.elastic.clients.elasticsearch.ElasticsearchAsyncClient;
 import co.elastic.clients.elasticsearch.ElasticsearchClient;
 import co.elastic.clients.elasticsearch.ElasticsearchTestServer;
 import co.elastic.clients.elasticsearch._types.ErrorResponse;
+import co.elastic.clients.elasticsearch._types.Refresh;
 import co.elastic.clients.elasticsearch._types.Script;
 import co.elastic.clients.elasticsearch._types.analysis.LimitTokenCountTokenFilter;
 import co.elastic.clients.elasticsearch._types.mapping.Property;
 import co.elastic.clients.elasticsearch._types.mapping.RuntimeField;
 import co.elastic.clients.elasticsearch._types.mapping.RuntimeFieldType;
 import co.elastic.clients.elasticsearch.cluster.ClusterStatsResponse;
+import co.elastic.clients.elasticsearch.core.GetResponse;
 import co.elastic.clients.elasticsearch.core.SearchRequest;
 import co.elastic.clients.elasticsearch.core.SearchResponse;
 import co.elastic.clients.elasticsearch.core.search.Suggester;
@@ -40,12 +43,15 @@ import co.elastic.clients.elasticsearch.model.ModelTestCase;
 import co.elastic.clients.elasticsearch.snapshot.RestoreResponse;
 import co.elastic.clients.json.JsonData;
 import co.elastic.clients.json.JsonpDeserializer;
+import co.elastic.clients.util.BinaryData;
+import co.elastic.clients.util.ContentType;
 import jakarta.json.stream.JsonParser;
 import org.junit.jupiter.api.Disabled;
 import org.junit.jupiter.api.Test;
 
 import java.io.InputStream;
 import java.io.StringReader;
+import java.nio.charset.StandardCharsets;
 
 /**
  * Test issues related to the API specifications.
@@ -53,6 +59,41 @@ import java.io.StringReader;
  * Depending on the feedback provided, this may involve either loading a JSON file or sending requests to an ES server.
  */
 public class SpecIssuesTest extends ModelTestCase {
+
+    @Test
+    public void i0575_asyncBinaryData() throws Exception {
+
+        ElasticsearchAsyncClient esAsyncClient = ElasticsearchTestServer.global().asyncClient();
+
+        String index = "binary-ingestion-test";
+        String id = "foo-bar";
+
+        BinaryData data = BinaryData.of(
+            ("{\"id\":\"27082ce1-d9ab-404e-a810-f2894640edf4\",\"firstName\":\"Jesse\",\"lastName\":\"Pinkman\",\"addresses\":" +
+                "[{\"street\":\"1001 Central Ave NE\",\"city\":\"Albuquerque\",\"state\":\"NM\",\"zip\":\"87106\"}]}").getBytes(),
+            ContentType.APPLICATION_JSON
+        );
+
+        esAsyncClient.index(i -> i
+            .index(index)
+            .id(id)
+            .document(data)
+            .refresh(Refresh.True)
+        ).get();
+
+        GetResponse<BinaryData> getResponse = esAsyncClient.get(g -> g
+                .index(index)
+                .id(id)
+            , BinaryData.class
+        ).get();
+
+        assertEquals(id, getResponse.id());
+        assertEquals(
+            "{\"id\":\"27082ce1-d9ab-404e-a810-f2894640edf4\",\"firstName\":\"Jesse\",\"lastName\":\"Pinkman\",\"addresses\":" +
+                "[{\"street\":\"1001 Central Ave NE\",\"city\":\"Albuquerque\",\"state\":\"NM\",\"zip\":\"87106\"}]}",
+            new String(getResponse.source().asByteBuffer().array(), StandardCharsets.UTF_8)
+        );
+    }
 
     @Test
     public void i0328_charFilter() throws Exception {

--- a/java-client/src/test/java/co/elastic/clients/json/JsonpUtilsTest.java
+++ b/java-client/src/test/java/co/elastic/clients/json/JsonpUtilsTest.java
@@ -25,13 +25,14 @@ import co.elastic.clients.elasticsearch.security.IndexPrivilege;
 import co.elastic.clients.elasticsearch.security.IndicesPrivileges;
 import co.elastic.clients.elasticsearch.security.RoleTemplateScript;
 import co.elastic.clients.elasticsearch.security.UserIndicesPrivileges;
-import co.elastic.clients.json.JsonpUtils;
 import co.elastic.clients.util.AllowForbiddenApis;
 import jakarta.json.JsonException;
 import jakarta.json.spi.JsonProvider;
 import jakarta.json.stream.JsonGenerator;
+import jakarta.json.stream.JsonParser;
 import org.junit.jupiter.api.Test;
 
+import java.io.StringReader;
 import java.io.StringWriter;
 import java.net.URL;
 import java.util.Collections;
@@ -209,6 +210,47 @@ public class JsonpUtilsTest extends ModelTestCase {
         }
     }
 
+    @Test
+    public void testCopy() {
+        // Tests round-tripping a json document that contains all event types and various kinds of nesting
+
+        String json = "{\n" +
+            "  \"p1\": \"str1\",\n" +
+            "  \"p2\": 42,\n" +
+            "  \"p3\": [\"str31\", \"str32\"],\n" +
+            "  \"p4\": {\n" +
+            "    \"p41\": \"str41\",\n" +
+            "    \"p42\": [\"str421\", \"str422\"],\n" +
+            "    \"p43\": {\n" +
+            "      \"p431\": \"str431\"\n" +
+            "    },\n" +
+            "    \"p44\": true,\n" +
+            "    \"p45\": false,\n" +
+            "    \"p46\": 3.14\n" +
+            "  },\n" +
+            "  \"p5\": [{\n" +
+            "    \"p51\": {\n" +
+            "      \"p511\": \"str511\"\n" +
+            "    }\n" +
+            "  }],\n" +
+            "  \"p6\": null\n" +
+            "}\n";
+
+        json = normalizeIndent(json);
+
+        JsonProvider provider = JsonpUtils.provider();
+
+        JsonParser parser = provider.createParser(new StringReader(json));
+        StringWriter sw = new StringWriter();
+        JsonGenerator generator = provider.createGenerator(sw);
+
+        JsonpUtils.copy(parser, generator);
+        parser.close();
+        generator.close();
+
+        assertEquals(json, sw.toString());
+    }
+
     private static String orNullHelper(Consumer<JsonGenerator> c) {
         StringWriter sw = new StringWriter();
         JsonGenerator generator = JsonpUtils.provider().createGenerator(sw);
@@ -220,5 +262,70 @@ public class JsonpUtilsTest extends ModelTestCase {
         generator.close();
 
         return sw.toString();
+    }
+
+    /**
+     * Normalizes the whitespace and indentation of a JSON string by parsing it and copying it to a string generator.
+     */
+    private static String normalizeIndent(String json) {
+        JsonParser parser = JsonpUtils.provider().createParser(new StringReader(json));
+        StringWriter sw = new StringWriter();
+        JsonGenerator generator = JsonpUtils.provider().createGenerator(sw);
+
+        copyAll(parser, generator);
+
+        parser.close();
+        generator.close();
+        return sw.toString();
+    }
+
+    private static void copyAll(JsonParser parser, JsonGenerator generator) {
+        while(parser.hasNext()) {
+            switch (parser.next()) {
+                case START_OBJECT:
+                    generator.writeStartObject();
+                    break;
+
+                case END_OBJECT:
+                    generator.writeEnd();
+                    break;
+
+                case START_ARRAY:
+                    generator.writeStartArray();
+                    break;
+
+                case END_ARRAY:
+                    generator.writeEnd();
+                    break;
+
+                case KEY_NAME:
+                    generator.writeKey(parser.getString());
+                    break;
+
+                case VALUE_STRING:
+                    generator.write(parser.getString());
+                    break;
+
+                case VALUE_NULL:
+                    generator.writeNull();
+                    break;
+
+                case VALUE_TRUE:
+                    generator.write(true);
+                    break;
+
+                case VALUE_FALSE:
+                    generator.write(false);
+                    break;
+
+                case VALUE_NUMBER:
+                    if (parser.isIntegralNumber()) {
+                        generator.write(parser.getLong());
+                    } else {
+                        generator.write(parser.getBigDecimal());
+                    }
+                    break;
+            }
+        }
     }
 }

